File: 1201-1300/1269-Number of Ways to Stay in the Same Place After Some Steps/1269-Number of Ways to Stay in the Same Place After Some Steps.py
from functools import lru_cache
class Solution:
    def numWays(self, steps: int, arrLen: int) -> int:
        MOD = 10 ** 9 + 7
        @lru_cache(None)
        def dfs(pos, steps):
            if pos == 0 and steps == 0:
                return 1
            if pos < 0 or pos >= arrLen or pos > steps:
                return 0
            steps -= 1
            return (dfs(pos - 1, steps) + dfs(pos, steps) + dfs(pos + 1, steps)) % MOD
        return dfs(0, steps)

File: 1201-1300/1269-Number of Ways to Stay in the Same Place After Some Steps/1269-Number of Ways to Stay in the Same Place After Some Steps.cpp
class Solution {
public:
    int numWays(int steps, int arrLen) {
        int sz = min(steps / 2 + 1, arrLen);
        vector<int> dp(sz);
        dp[0] = 1;
        int MOD = 1e9 + 7;
        for (int i = 0; i < steps; ++i) {
            vector<int> temp(sz);
            for (int j = 0; j < sz; ++j) {
                temp[j] = (((j > 0) ? dp[j - 1] : 0) +
                          (long)dp[j] + ((j < sz - 1) ? dp[j + 1] : 0)) % MOD;
            }
            swap(temp, dp);
        }
        
        return dp[0];
    }
};
